### PR TITLE
Distribute fixed width data

### PR DESCRIPTION
**Pull request by chrisrichardson**
Topology data is currently distributed as an `AdjacencyList`. This PR reduces memory usage and improves efficiency of mesh creation in parallel, by sending fixed-width data for topology (which is always the case).
Preparation for sending multiple cell types in a future PR.

### CODE REVIEW

File: cpp/dolfinx/graph/partition.h [hunk 1]
```diff
@@ -69,6 +69,25 @@ std::tuple<graph::AdjacencyList<std::int64_t>, std::vector<int>,
...
+/// @brief Distribute fixed-width nodes to destination ranks.
+///
+/// The global index of each node is assumed to be the local index plus
+/// the offset for this rank.
+///
+/// @param[in] comm MPI Communicator
+/// @param[in] list The list to distribute, as a flattened 2D row major array
+/// @param[in] width The width of the array
+/// @param[in] destinations Destination ranks for the ith node in the
+/// adjacency list. The first rank is the 'owner' of the node.
```
Comment (chrisrichardson) [inline cpp/dolfinx/graph/partition.h:81]:
```suggestion
///  list. The first rank is the 'owner' of the node.
```

File: cpp/dolfinx/graph/partition.h [hunk 2]
```diff
@@ -69,6 +69,25 @@ std::tuple<graph::AdjacencyList<std::int64_t>, std::vector<int>,
 distribute(MPI_Comm comm, const graph::AdjacencyList<std::int64_t>& list,
            const graph::AdjacencyList<std::int32_t>& destinations);
 
+/// @brief Distribute fixed-width nodes to destination ranks.
```
Comment (garth-wells) [inline cpp/dolfinx/graph/partition.h:72]: What is a 'fixed-width node`?

File: cpp/dolfinx/graph/partition.h [hunk 3]
```diff
@@ -69,6 +69,25 @@ std::tuple<graph::AdjacencyList<std::int64_t>, std::vector<int>,
...
+///
+/// The global index of each node is assumed to be the local index plus
+/// the offset for this rank.
+///
+/// @param[in] comm MPI Communicator
+/// @param[in] list The list to distribute, as a flattened 2D row major array
```
Comment (garth-wells) [inline cpp/dolfinx/graph/partition.h:78]: List of what? No need for the doc string to repeat the type, needs to explain what the argument is.

File: cpp/dolfinx/graph/partition.h [hunk 4]
```diff
@@ -69,6 +69,25 @@ std::tuple<graph::AdjacencyList<std::int64_t>, std::vector<int>,
...
+/// @param[in] width The width of the array
```
Comment (garth-wells) [inline cpp/dolfinx/graph/partition.h:79]: Use fewer word and more examples! Does list have shape (`list.size()/width, width)?

`width` is not a widely recognised term. Should it be `shape1`?

**Review by garth-wells — DISMISSED** [review]

**Comment by chrisrichardson** [comment]
Thanks for comments @garth-wells - sometimes writing the documentation is the hardest part.

**Pull request merged**